### PR TITLE
[Model] Add YOCO model implementation

### LINKED ISSUE #106: [RFC] Add YOCO models

### PR DESCRIPTION
**Pull request by Shomvel**
## Summary
This PR adds YOCO implementation as in https://arxiv.org/abs/2405.05254:
- Introduces the YOCO hybridization pattern of consecutive efficient attention layers (self-decoders), followed by a stack of full attention layers (cross-decoders) that share the KV cache projected from the last self-decoder layer (shared kv builder). This reduces the KV cache footprint and enables prefill to skip the cross-decoders (`skip_cross_decoder`). 
- Follows the official implementation by adding GatedRetention as a self-decoder option based on `simple_gla`, using the official `inv_freq` formulation, enabling interleaved rotary embeddings, and scaling the input embeddings.
- Adds extensions including GatedDeltaNet as a self-decoder option and configurable self-decoder layer counts (`num_self_decoder_layers`).

## Code Changes
- `fla/models/yoco`: Adds `YOCOConfig`, `YOCOModel`, and `YOCOForCausalLM`. `YOCOConfig` introduces `self_decoder_attn` and `cross_decoder_attn` dictionaries to configure the self-decoder and cross-decoder attention paths, and `YOCOModel` wires the split self-decoder / cross-decoder stack with configurable `num_self_decoder_layers`.
- `fla/layers/yoco_attn.py`: Adds `YOCOGatedRetention` as a self-decoder attention option, `YOCOStateRotaryEmbedding` to support YOCO-style `rope_inv_freq` and interleaved rotary embeddings, with rotary applied to `q` and `k` in self-decoder GatedRetention, to `k` only in the shared-KV builder, and to `q` only in cross-attention, `YOCOSharedKVBuilder` to build and cache the shared KV states for the cross-decoder, and `YOCOCrossAttention` to consume the shared KV states in the cross-decoder layers.

## Test Plan
- Added `tests/models/test_yoco.py` coverage for:
  - YOCO model construction across GatedRetention, GatedDeltaNet, and SWA self-decoder options.
  - Official-parity behaviors including YOCO-specific initialization, `rope_inv_freq='yoco'`, interleaved rotary usage in cross-decoder paths, and scaled token embeddings.
  - Decode/prefill behaviors including one-step cross attention, skipping cross-decoders during prefill, shared KV cache construction, cache defaults during training, gradient checkpointing backward smoke, and generation smoke tests.

## Breaking changes
None. This PR adds a new YOCO model/config path and experimental self-decoder options without changing existing model APIs.

### CODE REVIEW

File: fla/layers/yoco_attn.py [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+        **kwargs,
+    ) -> tuple[torch.Tensor, torch.Tensor | None, Cache | None]:
+        del output_attentions
+
+        if attention_mask is not None:
+            assert len(attention_mask.shape) == 2, (
+                "Expected attention_mask as a 0-1 matrix with shape [batch_size, seq_len] "
+                "for padding purposes (0 indicating padding). "
+                "Arbitrary attention masks of shape [batch_size, seq_len, seq_len] are not allowed."
+            )
```
[the remark this answers is not in the document]
Comment (Shomvel) [reply fla/layers/yoco_attn.py]: Good catch. I switched the three `attention_mask` shape guards in `fla/layers/yoco_attn.py` from `assert` to explicit `ValueError` checks in 3e5cd436.

File: fla/layers/yoco_attn.py [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+        self.num_heads = num_heads
+        self.num_kv_heads = self.num_heads if num_kv_heads is None else num_kv_heads
+        self.head_dim = self.hidden_size // self.num_heads
+        self.kv_dim = self.num_kv_heads * self.head_dim
+        self.qkv_bias = qkv_bias
+        self.rope_theta = rope_theta
+        self.max_position_embeddings = max_position_embeddings
+        self.layer_idx = layer_idx
+
+        self.kv_norm = (RMSNorm if fuse_norm else nn.RMSNorm)(self.hidden_size, eps=norm_eps)
```
[the remark this answers is not in the document]
Comment (Shomvel) [reply fla/layers/yoco_attn.py:251]: I’m keeping this as-is. The project already requires `torch>=2.7.0` in both `pyproject.toml` and `setup.py`, so `nn.RMSNorm` is available on all supported installs. The `(RMSNorm if fuse_norm else nn.RMSNorm)` pattern is also used across the existing model code, so I’d rather keep YOCO aligned with the repo-wide convention than special-case it here.

File: fla/layers/yoco_attn.py [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+            o = attn_decoding_one_step(q, shared_k, shared_v, cu_seqlens=cu_seqlens)
+        else:
+            o = parallel_attn(q, shared_k, shared_v, window_size=self.window_size, cu_seqlens=cu_seqlens)
+        if should_transform_output:
+            o = rearrange(o, 't b h d -> b t h d')
+        if pad_indices is not None:
+            o = pad_input(o.squeeze(0), pad_indices, batch_size, q_len)
+        o = o.reshape(batch_size, q_len, -1)
+        o = self.o_proj(o)
+        attentions = None if not output_attentions else None
```
[the remark this answers is not in the document]
Comment (Shomvel) [reply fla/layers/yoco_attn.py]: Agreed. Simplified this to a direct `attentions = None` assignment in 3e5cd436.

File: fla/models/yoco/modeling_yoco.py [hunk 4]
```diff
@@ -0,0 +1,694 @@
...
+    return hidden_states[batch_indices, last_token_offsets].unsqueeze(1)
+
+
+class YOCOSelfDecoderLayer(nn.Module):
+    def __init__(self, config: YOCOConfig, layer_idx: int):
+        super().__init__()
+
+        self.config = config
+        self.layer_idx = layer_idx
+        self.attn_norm = (RMSNorm if config.fuse_norm else nn.RMSNorm)(config.hidden_size, eps=config.norm_eps)
```
[the remark this answers is not in the document]
Comment (Shomvel) [reply fla/models/yoco/modeling_yoco.py:61]: Same rationale here: the repo minimum is `torch>=2.7.0`, and the existing model implementations consistently use `(RMSNorm if config.fuse_norm else nn.RMSNorm)`. I’m keeping YOCO on that same convention instead of diverging within a single model family.

File: fla/models/yoco/modeling_yoco.py [hunk 5]
```diff
@@ -0,0 +1,694 @@
...
+        past_key_values: Cache | list[torch.FloatTensor] | None = None,
+        use_cache: bool | None = None,
+        output_attentions: bool | None = None,
+        output_hidden_states: bool | None = None,
+        return_dict: bool | None = None,
+        skip_cross_decoder: bool = False,
+        **kwargs: Unpack[dict]
+    ) -> tuple | BaseModelOutputWithPast:
+        if output_attentions:
+            warnings.warn("`YOCOModel` does not `output_attentions` now, setting it to `False`.")
```
[the remark this answers is not in the document]
Comment (Shomvel) [reply fla/models/yoco/modeling_yoco.py]: Agreed. Switched this to `logger.warning_once(...)` in 3e5cd436 so repeated forward calls do not spam logs.

**Comment by yzhangcs** [comment]
@Shomvel hey any update on this pr?

**Comment by yzhangcs** [comment]
and could you adapt it to recent attnres merges
